README.md
# mod-courses

Copyright (C) 2019 The Open Library Foundation

This software is distributed under the terms of the Apache License,
Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

## Introduction

Backend CRUD storage module to support the Course Reserves functionality

## Additional information

Other [modules](https://dev.folio.org/source-code/#client-side).

See project [MODCR](https://issues.folio.org/browse/MODCR)
at the [FOLIO issue tracker](https://dev.folio.org/guidelines/issue-tracker).

Other FOLIO Developer documentation is at [dev.folio.org](https://dev.folio.org/)


NEWS.md
## 1.2.3 2021-06-19
* Correct issue with temporaryLocationId removal (MODCR-58)
* Update interface versions for inventory (MODCR-59)
* Correct issue with potential CQL injection (MODCR-62)

## 1.2.2 2021-04-26
* Upgrade to RMB 32.2.2 (MODCR-58)
* Add module permissions to PUT for requests
* Clean up technical debt (MODCR-54, MODCR-55)

## 1.2.1 2021-04-23
* Upgrade to RMB 32.2.1 (MODCR-57)
* Fix issue with updating temporaryLocationId to null value (MODCR-56)

## 1.2.0 2021-03-10
* Upgrade to RMB 32.2.0

## 1.1.2 2021-01-28
* Add ForeignKey entry for Coursetype into Courselistings table (MODCR-50)

## 1.1.1 2020-11-06
* Update to RMB v31.1.5, Vertx 3.9.4 (MODCR-46)

## 1.1.0 2020-10-05
* Update to RMB v31, JDK 11 (MODCR-43)

## 1.0.7 2020-09-15
* Allow tenant init even if load sample fails (MODCR-41)

## 1.0.6 2020-09-10
* Fix bug with tenant init parameters not being parsed (MODCR-42)

## 1.0.5 2020-06-03
* Fix bug with instructorObject cache not getting refreshed on PUT (MODCR-39)

## 1.0.4.2020-06-02
* Update to RMB 30.0.1 (MODCR-38)

## 1.0.3 2020-04-27
* Write changes in temporaryLoanTypeId back to item record on POST/PUT for reserves

## 1.0.2 2020-04-07
* Add missing module permissions for location and servicepoint lookup

## 1.0.1 2020-04-06

* Populate temporary location field from CourseListing (MODCR-17,MODCR-31)
* Properly pull call number from Item/Holdings records (MODCR-29,MODCR-30)
* Update RAML capitalization to work on MacOS (MODCR-2,  MODCR-25)
* Fix bug with deletion of a Reserve with no corresponding Item (MODCR-22)
* Fix parameters to loadSample on tenant init (MODCR-23, MODCR-35)
* Change tests to use shared HttpClient rather than new (MODCR-26)
* Fix population for electronicAccess fields with fallback (MODCR-28)
* Scrub fields marked as dynamic from POST/PUT (MODCR-34)
* Fix issue with field expansion for lists of Reserves (MODCR-36)


## 1.0.0 2020-03-13

* Initial release


src/test/java/UnitTest/UnitTest.java
package UnitTest;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import org.folio.coursereserves.util.CRUtil;
import org.folio.coursereserves.util.PopulateMapping;
import org.folio.coursereserves.util.PopulateMapping.ImportType;
import org.folio.rest.jaxrs.model.LocationObject;
import org.folio.rest.jaxrs.model.LocationObject;
import org.folio.rest.jaxrs.model.TemporaryLocationObject;
import static org.junit.Assert.assertTrue;
import org.junit.Test;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertTrue;

public class UnitTest {
  
  @Test
  public void testPojoFromJson() throws Exception {
    LocationObject locationObject = new LocationObject();
    String name = "Front Desk";
    String id = UUID.randomUUID().toString();
    JsonObject json = new JsonObject();
    json.put("id", id);
    json.put("name", name);
    json.put("isActive", Boolean.TRUE);
    JsonArray servicePointIds = new JsonArray();
    servicePointIds.add(UUID.randomUUID().toString());
    servicePointIds.add(UUID.randomUUID().toString());
    servicePointIds.add(UUID.randomUUID().toString());
    json.put("servicePointIds", servicePointIds);
    List<PopulateMapping> mapList = new ArrayList<>();
    mapList.add(new PopulateMapping("id"));
    mapList.add(new PopulateMapping("name"));
    mapList.add(new PopulateMapping("isActive", ImportType.BOOLEAN));
    mapList.add(new PopulateMapping("servicePointIds", ImportType.STRINGLIST));
    CRUtil.populatePojoFromJson(locationObject, json, mapList);
    assertTrue(locationObject.getName().equals(name));
    assertTrue(locationObject.getId().equals(id));
    assertTrue(locationObject.getIsActive());
    assertTrue(locationObject.getServicePointIds().size() == 3);
  }
  
  @Test
  public void testCopyFields() {
    TemporaryLocationObject tempLocationObject = new TemporaryLocationObject();
    LocationObject locationObject = new LocationObject();
    locationObject.setId(UUID.randomUUID().toString());
    locationObject.setName("Big Library");
    locationObject.setIsActive(Boolean.TRUE);
    CRUtil.copyFields(tempLocationObject, locationObject);
    assertTrue(tempLocationObject.getId().equals(locationObject.getId()));
    assertTrue(tempLocationObject.getName().equals(locationObject.getName()));
    assertTrue(tempLocationObject.getIsActive().equals(locationObject.getIsActive()));
  }


  
}


src/test/java/CourseAPITest/TestUtil.java
package CourseAPITest;

import io.vertx.core.Future;
import io.vertx.core.MultiMap;
import io.vertx.core.Promise;
import io.vertx.core.Vertx;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpClient;
import io.vertx.core.http.HttpClientRequest;
import io.vertx.core.http.HttpClientResponse;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import io.vertx.ext.web.client.HttpRequest;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.client.WebClient;
import java.util.Map;


public class TestUtil {
  private static HttpClient httpClient = null;
  private static final Logger logger = LoggerFactory.getLogger(TestUtil.class.getClass()
      .getName());

  static class WrappedResponse {
    private String explanation;
    private int code;
    private String body;
    private JsonObject json;
    private HttpResponse<Buffer> response;

    public WrappedResponse(String explanation, int code, String body,
        HttpResponse<Buffer> response) {
      this.explanation = explanation;
      this.code = code;
      this.body = body;
      this.response = response;
      try {
        json = new JsonObject(body);
      } catch(Exception e) {
        json = null;
      }
    }

    public String getExplanation() {
      return explanation;
    }

    public int getCode() {
      return code;
    }

    public String getBody() {
      return body;
    }

    public HttpResponse<Buffer> getResponse() {
      return response;
    }

    public JsonObject getJson() {
      return json;
    }

  }

  private static HttpClient getHttpClient(Vertx vertx) {
    if(httpClient == null) {
      httpClient = vertx.createHttpClient();
    }
    return httpClient;
  }

  public static Future<WrappedResponse> doOkapiRequest(Vertx vertx,
      String requestPath, HttpMethod method, Map<String, String> okapiHeaders,
      Map<String, String> extraHeaders, String payload,
      Integer expectedCode, String explanation) {
    WebClient client = WebClient.create(vertx);
    MultiMap headers = MultiMap.caseInsensitiveMultiMap();
    MultiMap originalHeaders = MultiMap.caseInsensitiveMultiMap();
    originalHeaders.setAll(okapiHeaders);
    String okapiUrl = originalHeaders.get("x-okapi-url");
    if(okapiUrl == null) {
      return Future.failedFuture("No okapi URL found in headers");
    }
    String requestUrl = okapiUrl + requestPath;
    if(originalHeaders.contains("x-okapi-token")) {
      headers.add("x-okapi-token", originalHeaders.get("x-okapi-token"));
    }
    if(originalHeaders.contains("x-okapi-tenant")) {
      headers.add("x-okapi-tenant", originalHeaders.get("x-okapi-tenant"));
    }
    headers.add("content-type", "application/json");
    headers.add("accept", "application/json");
    if(extraHeaders != null) {
      for(Map.Entry<String, String> entry : extraHeaders.entrySet()) {
        headers.add(entry.getKey(), entry.getValue());
      }
    }
    HttpRequest<Buffer> request = client.requestAbs(method, requestUrl);
    for(Map.Entry entry : headers.entries()) {
      String key = (String)entry.getKey();
      String value = (String)entry.getValue();
      if( key != null && value != null) {
        request.putHeader(key, value);
      }
    }
    return wrapRequestResponse(request, payload, requestUrl, method, expectedCode, explanation);
  }

  public static Future<WrappedResponse> doRequest(Vertx vertx, String url,
          HttpMethod method, MultiMap headers, String payload,
          Integer expectedCode, String explanation) {

    boolean addPayLoad = false;
    //HttpClient client = vertx.createHttpClient();
    WebClient client = WebClient.create(vertx);
    HttpRequest<Buffer> request = client.requestAbs(method, url);
    //Add standard headers
    request.putHeader("X-Okapi-Tenant", "diku")
            .putHeader("content-type", "application/json")
            .putHeader("accept", "application/json");
    if(headers != null) {
      for(Map.Entry entry : headers.entries()) {
        request.putHeader((String)entry.getKey(), (String)entry.getValue());
        System.out.println(String.format("Adding header '%s' with value '%s'",
            (String)entry.getKey(), (String)entry.getValue()));
      }
    }
    /*
    //standard exception handler
    request.exceptionHandler(e -> { future.fail(e); });
    request.handler( req -> {
      req.bodyHandler(buf -> {
        String explainString = "(no explanation)";
        if(explanation != null) { explainString = explanation; }
        if(expectedCode != null && expectedCode != req.statusCode()) {
          future.fail(method.toString() + " to " + url + " failed. Expected status code "
                  + expectedCode + ", got status code " + req.statusCode() + ": "
                  + buf.toString() + " | " + explainString);
        } else {
          System.out.println("Got status code " + req.statusCode() + " with payload of: " + buf.toString() + " | " + explainString);
          WrappedResponse wr = new WrappedResponse(explanation, req.statusCode(), buf.toString(), req);
          future.complete(wr);
        }
      });
    });
    System.out.println("Sending " + method.toString() + " request to url '"+
              url + " with payload: " + payload + "'\n");
    if(method == HttpMethod.PUT || method == HttpMethod.POST) {
      request.end(payload);
    } else {
      request.end();
    }
    return future;
    */
    return wrapRequestResponse(request, payload, url, method, expectedCode, explanation);
  }

  /* HttpClientRequest should have headers set before calling */
  private static Future<WrappedResponse> wrapRequestResponse(HttpRequest<Buffer> request,
      String payload, String uri, HttpMethod method, Integer expectedCode, String explanation) {
    Promise<WrappedResponse> promise = Promise.promise();
    Future<HttpResponse<Buffer>> sentRequestFuture;

    if( method == HttpMethod.PUT || method == HttpMethod.POST ) {
      sentRequestFuture = request.sendBuffer(Buffer.buffer(payload));
    } else {
      sentRequestFuture = request.send();
    }
    sentRequestFuture.onComplete(res -> {
      if(res.failed()) {
        promise.fail(res.cause());
      } else {
        try {
          HttpResponse<Buffer> result = res.result();
          String responseString = result.bodyAsString();
          String explainString = (explanation != null) ? explanation : "(no explanation)";
          if(expectedCode != null && expectedCode != result.statusCode()) {
            String message = method.toString() + " to " + uri
                  + " failed. Expected status code "
                  + expectedCode + ", got status code " + result.statusCode() + ": "
                  + responseString + " | " + explainString;
            promise.fail(message);
            logger.error(message);
          } else {
            logger.info("Got status code " + result.statusCode() + " with payload of: "
                + responseString + " | " + explainString);
            WrappedResponse wr = new WrappedResponse(explanation, result.statusCode(),
                responseString, result);
            promise.complete(wr);
          }
        } catch(Exception e) {
          promise.fail(e);
        }
      }
    });
    
    logger.info("Sending " + method.toString() + " request to url '"+
              uri + " with payload: " + payload + "'\n");

    return promise.future();

  }
}



ramls/tmp/new-model-sample-data.js
const data = {
  terms: [
    {
      id: 14,
      name: "Summer 2019",
      startDate: "2019-04-03",
      endDate: "2019-07-31"
    },
    {
      id: 15,
      name: "Autumn 2019",
      startDate: "2019-09-01",
      endDate: "2019-12-18"
    }
  ],
  courseListings: [
    {
      id: 21,
      term: 15,
    },
    {
      id: 22,
      term: 14,
    }
  ],
  courses: [
    {
      name: "History 201",
      department: "History",
      courseListingId: 21
    },
    {
      name: "Anthropology 101",
      department: "Anthropology",
      courseListingId: 21
    },
    {
      name: "Calculus 201",
      department: "Mathematics",
      courseListingId: 22
    }
  ],
  reserves: [
    {
      itemId: 35,
      courseListingId: 21,
      copiedData: { title: "Tudor History" },
    },
    {
      itemId: 36,
      courseListingId: 21,
      copiedData: { title: "History of WWII" },
    },
    {
      itemId: 37,
      courseListingId: 22,
      copiedData: { title: "Integration for Dummies" },
    }
  ]
};

console.log(data);


src/main/java/org/folio/rest/impl/CoursesTenantAPI.java
package org.folio.rest.impl;

import io.vertx.core.AsyncResult;
import io.vertx.core.Context;
import io.vertx.core.Handler;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import java.util.List;
import java.util.Map;
import org.folio.rest.jaxrs.model.TenantAttributes;
import javax.ws.rs.core.Response;
import org.folio.rest.tools.utils.TenantLoading;
import org.folio.coursereserves.util.Util;
import org.folio.rest.jaxrs.model.Parameter;



public class CoursesTenantAPI extends TenantAPI {

  public static final String SAMPLE_DATA_COURSELISTING = "c03bcba3-a6a0-4251-b316-0631bb2e6f21";
  public static final Logger logger = LoggerFactory.getLogger(CoursesTenantAPI.class);
  protected static final String PARAMETER_LOAD_SAMPLE = "loadSample";

  @Override
  public void postTenant(TenantAttributes tenantAttributes, Map<String, String> headers,
      Handler<AsyncResult<Response>> handler, Context context) {
    logger.info("Calling overridden postTenant method");
    Boolean loadSample;
    List<Parameter> parameterList = tenantAttributes.getParameters();
    Boolean loadSampleCandidate = null;
    for(Parameter parameter : parameterList) {      
      if(PARAMETER_LOAD_SAMPLE.equals(parameter.getKey())) {
        loadSampleCandidate = Boolean.parseBoolean(parameter.getValue());
      }      
    }
    loadSample = loadSampleCandidate != null ? loadSampleCandidate : false;
    super.postTenantSync(tenantAttributes, headers, res -> {
      if(res.failed()) {
        logger.error("Unable to load tenant: " + res.cause().getMessage());
        handler.handle(res);
      } else if(Boolean.TRUE.equals(loadSample)) {
        TenantLoading tenantLoading = new TenantLoading();
        tenantLoading
            .withKey(PARAMETER_LOAD_SAMPLE).withLead("sample-data").withPostOnly()
              .add("terms", "coursereserves/terms")
              .add("copyrightstatuses", "coursereserves/copyrightstatuses")
              .add("departments", "coursereserves/departments")
              .add("processingstatuses", "coursereserves/processingstatuses")
              .add("coursetypes", "coursereserves/coursetypes")
              .add("courselistings", "coursereserves/courselistings")
              .add("courses", "coursereserves/courses")
              .add("instructors", "coursereserves/courselistings/"+SAMPLE_DATA_COURSELISTING+"/instructors")
              .add("reserves", "coursereserves/reserves");
        tenantLoading.perform(tenantAttributes, headers, context.owner(), 
            performRes -> {
          if(performRes.failed()) {
            Util.logAndSaveError(performRes.cause(), logger);
          }
          // We're gonna go ahead and return success even if the sample load fails
          handler.handle(res); //This should allow the proper response for upgrade or install
          
        });
            
      } else {
        handler.handle(res);
      }
    },
    context);
  }
  
}


util/load_data/load_data.py
import re
import csv
import requests

standard_headers = {
    "Content-Type": "application/json",
    "Accept": "application/json"
}

def make_headers(token, tenant):
    headers = standard_headers.copy()
    headers["X-Okapi-Token"] = token
    headers["X-Okapi-Tenant"] = tenant
    return headers

def get_courselisting_by_external_id(okapi_url, tenant, token, external_id):
    courselisting_url = "%s/%s?query=externalId=%s" % (okapi_url, "coursereserves/courselistings",
            external_id)
    response = requests.get(courselisting_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get courselisting with externalId %s, %s" %
                (external_id, response.text))
    if len(response.json()["courseListings"]) < 1:
            return None
    return response.json()["courseListings"][0]["id"]

def post_courselisting(okapi_url, tenant, token, external_id, term_id):
    courselisting_url = "%s/%s" % (okapi_url, "coursereserves/courselistings")
    payload = { "externalId" : external_id, "termId" : term_id }
    response = requests.post(courselisting_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to create courselisting with payload %s: %s" %
                (payload, response.text))
    return response.json()["id"]

def get_department_by_name(okapi_url, tenant, token, name):
    department_url = "%s/%s?query=name=%s" % (okapi_url, "coursereserves/departments", name)
    response = requests.get(department_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get department with name %s, %s" %
                ( name, response.text ))
    if len(response.json()["departments"]) < 1:
        return None
    return response.json()["departments"][0]["id"]

def post_department(okapi_url, tenant, token, name):
    department_url = "%s/%s" % (okapi_url, "coursereserves/departments")
    payload = { "name" : name }
    response = requests.post(department_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to add department with payload %s: %s" %
                (payload, response.text))
    return response.json()["id"]
    
def get_instructor_by_courselisting_and_name(okapi_url, tenant,
        token, courselisting_id, name):
    instructor_url = "%s/%s/%s/instructors?query=name=%s" % (okapi_url,\
            "coursereserves/courselistings", courselisting_id, name)
    response = requests.get(instructor_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get instructor from url %s with name %s, %s" %
                ( instructor_url, name, response.text ))
    if len(response.json()["instructors"]) < 1:
        return None
    return response.json()["instructors"][0]["id"]

def post_instructor(okapi_url, tenant, token, courselisting_id, name):
    instructor_url = "%s/%s/%s/instructors" %\
            (okapi_url, "coursereserves/courselistings", courselisting_id)
    payload = { "name" : name, "courseListingId": courselisting_id }
    response = requests.post(instructor_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to add instructor to url %s with payload %s: %s" %
                (instructor_url, payload, response.text))
    return response.json()["id"]

def get_reserve_by_item_id(okapi_url, tenant, token, item_id):
    reserve_url = "%s/%s?query=itemId=%s" % (okapi_url, "coursereserves/reserves", item_id)
    response = requests.get(reserve_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get reserves with url %s, %s" %\
                (reserve_url, response.text))
    if len(response.json()["reserves"]) < 1:
        return None

    return response.json()["reserves"][0]["id"]

def post_reserve(okapi_url, tenant, token, courselisting_id, item_id):
    reserve_url = "%s/%s" % (okapi_url, "coursereserves/reserves")
    payload = { "courseListingId" : courselisting_id, "itemId" : item_id }
    response = requests.post(reserve_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to add reserve to url %s with payload %s: %s" %
                (reserve_url, payload, response.text))
    return response.json()["id"]

def post_department(okapi_url, tenant, token, name):
    department_url = "%s/%s" % (okapi_url, "coursereserves/departments")
    payload = { "name" : name }
    response = requests.post(department_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to add department to url %s with payload %s: %s" %
                (department_url, payload, response.text))
    return response.json()["id"]

def get_department_by_name(okapi_url, tenant, token, name):
    department_url = "%s/%s?query=name=%s" % (okapi_url, "coursereserves/departments", name)
    response = requests.get(department_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get department with url %s, %s" %\
                (department_url, response.text))
    if len(response.json()["departments"]) < 1:
        return None

    return response.json()["departments"][0]["id"]

def lookup_user_by_first_last_location(okapi_url, tenant, token, first_name,\
        last_name, location):
    user_url = "%s/%s?query=personal.lastName=%s+AND+personal.firstName=%s" % (okapi_url, "users",\
            last_name, first_name)
    response = requests.get(user_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get users with url %s, %s" %\
                (user_url, response.text))
    if len(response.json()["users"]) < 1:
        #print("No users returned for URL %s" % user_url)
        return None

    users = response.json()["users"]
    for user in users:
        user_location = None
        try:
            user_location = user["personal"]["addresses"][0]["city"]
        except Exception as e:
            print("Unable to get location: %s" % e)
        if user_location and user_location.upper() == location.upper():
            return user["id"]
        else:
            print("%s and %s don't match for locations" % (location, user_location))
    return None

def lookup_item_by_former_id(okapi_url, tenant, token, former_id):
    item_url = "%s/%s?query=formerIds=%s" % (okapi_url, "item-storage/items", former_id)
    response = requests.get(item_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get items with url %s, %s" %\
                (item_url, response.text))
    if len(response.json()["items"]) < 1:
        print("No items returned for url %s" % item_url)
        return None

    return response.json()["items"][0]["id"]

def post_course(okapi_url, tenant, token, courselisting_id, department_id, name, description):
    course_url = "%s/%s" % (okapi_url, "coursereserves/courses")
    payload = { "courseListingId": courselisting_id, "name" : name,\
            "description" : description, "departmentId": department_id }
    response = requests.post(course_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoaderException("Unable to add course to url %s with payload %s: %s" %
                (course_url, payload, response.text))

    return response.json()["id"]

def get_course_by_name(okapi_url, tenant, token, name):
    course_url = "%s/%s?query=name=%s" % (okapi_url, "coursereserves/courses", name)
    response = requests.get(course_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get course with url %s, %s" %\
                (course_url, response.text))
    if len(response.json()["courses"]) < 1:
        return None

    return response.json()["courses"][0]["id"]

def post_term(okapi_url, tenant, token, name, start_date, end_date):
    term_url = "%s/%s" % (okapi_url, "coursereserves/terms")
    payload = { "name" : name, "startDate" : start_date, "endDate" : end_date }
    response = requests.post(term_url, headers=make_headers(token, tenant),
            json=payload)
    if response.status_code is not 201:
        raise LoadException("Unable to add course to url %s with payload %s: %s" %
                (term_url, payload, response.text))
    return response.json()["id"]

def get_term_by_name(okapi_url, tenant, token, name):
    term_url = "%s/%s?query=name=%s" % ( okapi_url, "coursereserves/terms", name)
    response = requests.get(term_url, headers=make_headers(token, tenant))
    if response.status_code is not 200:
        raise LoaderException("Unable to get term with url %s: %s" %\
                (term_url, response.text))
    if len(response.json()["terms"]) < 1:
        return None

    return response.json()["terms"][0]["id"]
    

class LoaderException(Exception):
    pass

class Loader:
    def __init__(self, okapi_url, tenant, term_name):
        self.okapi_url = okapi_url
        self.tenant = tenant
        self.term_name = term_name
        self.token = None
        self.courselisting_dict = {}
        self.department_dict = {}
        self.instructor_dict = {}
        self.reserve_dict = {}
        self.term_dict = {}
    
    def login(self, username, password):
        login_url = self.okapi_url + "/authn/login"
        payload = { "username" : username, "password" : password }
        response = requests.post(login_url, headers=make_headers(None, self.tenant),
                json=payload)
        if response.status_code is not 201:
            raise LoaderException("Unable to login: %s" % response.text)
        self.token = response.headers['x-okapi-token']

    def get_term(self, term_name):
        if term_name in self.term_dict:
            return self.term_dict[term_name]
        term_id = get_term_by_name(self.okapi_url, self.tenant, self.token, term_name)
        if term_id is None:
            start_date = '2020-01-09T00:00:00+0000'
            end_date = '2020-09-15T00:00:00+0000'
            term_id = post_term(self.okapi_url, self.tenant, self.token, term_name,
                    start_date, end_date)
            self.term_dict[term_name] = term_id
        return term_id

    def get_courselisting(self, external_id):
        if external_id in courselisting_dict:
            return courselisting_dict[external_id]
        courselisting = get_courselisting_by_external_id(self.okapi_url, self.tenant,
                self.token, external_id)
        return courselisting

    def get_department(self, name):
        if name in self.department_dict:
            return self.department_dict[name]
        department_id = get_department_by_name(self.okapi_url, self.tenant, self.token, name)
        if department_id is None:
            department_id = post_department(self.okapi_url, self.tenant, self.token,
                    name)
            self.department_dict[name] = department_id
        return department_id

    def get_instructor(self, courselisting_id, name):
        key = "%s_%s" % (courselisting_id, name)
        if key in self.instructor_dict:
            return self.instructor_dict[key]
        instructor_id = get_instructor_by_courselisting_and_name(self.okapi_url,
                self.tenant, self.token, courselisting_id, name)
        if instructor_id is None:
            instructor_id = post_instructor(self.okapi_url, self.tenant,
                    self.token, courselisting_id, name)
            self.instructor_dict[key] = instructor_id
        return instructor_id

    def get_reserve(self, item_id):
        pass

    def extract_course_parts(self, course):
        pass

    def get_item_list(self, item_string):
        pass

    def sanitize_instructor_name(self, name):
        pattern = r"(.+),\s?(.+?)\s*\((.+)\)"
        res = re.match(pattern, name)
        if not res:
            raise LoaderException("Unable to match input '%s' with pattern '%s'" %\
                    (name, pattern))
        return (res.group(1), res.group(2), res.group(3))

    def sanitize_course_name(self, name):
        pattern = r"([a-zA-Z]+)\s+(\d+[a-zA-Z]*?)\s+-\s+(.+)"
        res = re.match(pattern, name)
        if not res:
            raise LoaderException("Unable to match input '%s' with pattern '%s'" %\
                    (name, pattern))
        return (res.group(1), res.group(2), res.group(3))

    def read_lines_from_file(self, handle):
        reader = csv.reader(handle, delimiter='|')
        term_id = self.get_term(self.term_name)
        next(reader) #skip header
        for row in reader:
            try:
                print("Create courselisting with external id '%s'" % row[0])
                courselisting_id = post_courselisting(self.okapi_url, self.tenant,\
                        self.token, row[0], term_id)
                course_parts = self.sanitize_course_name(row[1])
                print("Find or create department with name '%s'" % course_parts[0])
                department_id = self.get_department(course_parts[0])
                course_name = "%s %s" % ( course_parts[0], course_parts[1] )
                course_description = course_parts[2]
                print("Create course belonging to courselisting '%s' with name '%s'" %\
                        (row[0], course_name))
                course_id = post_course(self.okapi_url, self.tenant, self.token,
                        courselisting_id, department_id, course_name, course_description)
                instructor_parts = self.sanitize_instructor_name(row[2])
                print("Find or create instructor with courselisting '%s' and name '%s'" %\
                        (row[0], instructor_parts[0]))
                instructor_id = self.get_instructor(courselisting_id, instructor_parts[0])
                item_id_list = row[3].split('%')
                for n in range(len(item_id_list)):
                    item_id_list[n] = item_id_list[n].strip('"')
                print("Create reserves for the following items: %s" % item_id_list)
                for item in item_id_list:
                    print("Creating reserve for item %s" % item)
                    post_reserve(self.okapi_url, self.tenant, self.token, courselisting_id,
                            item)
            except Exception as e:
                #print("Unable to process row %s: %s" % (row, e))
                raise e

    def test_lookups_from_file(self, handle):
        reader = csv.reader(handle, delimiter='|')
        term_id = self.get_term(self.term_name)
        next(reader)
        for row in reader:
            try:
                instructor_parts = self.sanitize_instructor_name(row[2])
                user_id = lookup_user_by_first_last_location(self.okapi_url,\
                        self.tenant, self.token, instructor_parts[1],\
                        instructor_parts[0], instructor_parts[2])
                print("Returned user_id for %s, %s (%s) is %s" %\
                        (instructor_parts[0], instructor_parts[1], instructor_parts[2], user_id))
                item_id_list = [ a.strip('"') for a in row[3].split('%') if a ]
                for item in item_id_list:
                    former_id = item.split("::a",1)[0]
                    item_id = lookup_item_by_former_id(self.okapi_url, self.tenant,\
                            self.token, former_id)
                    print("Got id %s for former id %s" % ( item_id, former_id))
            except Exception as e:
                #raise e
                print("Unable to process row %s: %s" % (row, e))







